lms/djangoapps/courseware/grades.py

```diff
@@ -358,88 +414,95 @@ def _grade(student, request, course, keep_raw_scores, field_data_cache, scores_c
 
+        max_scores_cache.push_to_remote()
+
+    return grade_summary
+
+
+def _calculate_totaled_scores(
+        student,
+        grading_context_result,
+        max_scores_cache,
+        submissions_scores,
+        scores_client,
+        keep_raw_scores,
+):
+    """
+    Returns the totaled scores, which can be passed to the grader.
+    """
+    raw_scores = []
     totaled_scores = {}
-    # This next complicated loop is just to collect the totaled_scores, which is
-    # passed to the grader
-    for section_format, sections in grading_context['graded_sections'].iteritems():
+    for section_format, sections in grading_context_result['all_graded_sections'].iteritems():
         format_scores = []
-        for section in sections:
-            section_descriptor = section['section_descriptor']
-            section_name = section_descriptor.display_name_with_default_escaped
+        for section_info in sections:
+            section = section_info['section_block']
+            section_name = block_metadata_utils.display_name_with_default(section)
 
             with outer_atomic():
-                # some problems have state that is updated independently of interaction
-                # with the LMS, so they need to always be scored. (E.g. combinedopenended ORA1)
-                # TODO This block is causing extra savepoints to be fired that are empty because no queries are executed
-                # during the loop. When refactoring this code please keep this outer_atomic call in mind and ensure we
-                # are not making unnecessary database queries.
-                should_grade_section = any(
-                    descriptor.always_recalculate_grades for descriptor in section['xmoduledescriptors']
-                )
-
-                # If there are no problems that always have to be regraded, check to
+                # Check to
                 # see if any of our locations are in the scores from the submissions
                 # API. If scores exist, we have to calculate grades for this section.
-                if not should_grade_section:
-                    should_grade_section = any(
-                        descriptor.location.to_deprecated_string() in submissions_scores
-                        for descriptor in section['xmoduledescriptors']
-                    )
+                should_grade_section = any(
+                    unicode(descendant.location) in submissions_scores
+                    for descendant in section_info['scored_descendants']
+                )
 
                 if not should_grade_section:
                     should_grade_section = any(
-                        descriptor.location in scores_client
-                        for descriptor in section['xmoduledescriptors']
+                        descendant.location in scores_client
+                        for descendant in section_info['scored_descendants']
                     )
```

Comment:
I think you could assign should_grade_section in a single `or` statement, since the `or` operator will only evaluate the second term if the first returns `False`.